### PR TITLE
feat: local evaluation cohorts support

### PR DESCRIPTION
**Pull request by bgiori**

### CODE REVIEW

File: src/main/kotlin/LocalEvaluationClient.kt [hunk 1]
```diff
@@ -67,15 +94,108 @@ class LocalEvaluationClient internal constructor(
...
+    } else {
+        null
+    }
+}
+
+private fun getServerUrl(config: LocalEvaluationConfig): HttpUrl {
+    return if (config.serverZone == LocalEvaluationConfig.Defaults.SERVER_ZONE) {
+        config.serverUrl.toHttpUrl()
+    } else {
+        when (config.serverZone) {
```
Comment (tyiuhc) [inline src/main/kotlin/LocalEvaluationClient.kt:166]: So if an EU user wants to set a custom URL, they should skip setting serverZone?

File: src/main/kotlin/deployment/DeploymentRunner.kt [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            val deletedCohortIds = storageCohortIds - flagCohortIds
+            for (deletedCohortId in deletedCohortIds) {
+                cohortStorage.deleteCohort(deletedCohortId)
+            }
+        }
+        Logger.d("Refreshed ${flagConfigs.size} flag configs.")
+    }
+
+    private fun getCohortPollingInterval(): Long {
+        if (config.cohortSyncConfig == null || config.cohortSyncConfig.cohortPollingIntervalMillis < 60000) {
```
Comment (tyiuhc) [inline src/main/kotlin/deployment/DeploymentRunner.kt:135]: nit: can make 60000 a constant

File: src/main/kotlin/RemoteEvaluationConfig.kt [hunk 3]
```diff
@@ -89,6 +103,8 @@ class RemoteEvaluationConfig internal constructor(
 
         private var debug = Defaults.DEBUG
         private var serverUrl = Defaults.SERVER_URL
+        private var cohortServerUrl = Defaults.COHORT_SERVER_URL
```
Comment (tyiuhc) [inline src/main/kotlin/RemoteEvaluationConfig.kt:106]: is this used in remote evaluation client?

File: src/main/kotlin/LocalEvaluationConfig.kt [hunk 4]
```diff
@@ -39,6 +51,16 @@ class LocalEvaluationConfig internal constructor(
          */
         const val SERVER_URL = "https://api.lab.amplitude.com/"
 
+        /**
+         * "https://api.lab.amplitude.com/"
```
Comment (tyiuhc) [inline src/main/kotlin/LocalEvaluationConfig.kt:55]: fix comment

File: src/main/kotlin/LocalEvaluationClient.kt [hunk 5]
```diff
@@ -67,15 +94,108 @@ class LocalEvaluationClient internal constructor(
...
         val variants = evaluationResults.toVariants()
         assignmentService?.track(Assignment(user, variants))
         return variants
     }
+
+    private fun enrichUser(user: ExperimentUser, flagConfigs: List<EvaluationFlag>): ExperimentUser {
+        val groupedCohortIds = flagConfigs.getGroupedCohortIds()
+        if (cohortStorage == null) {
+            if (groupedCohortIds.isNotEmpty()) {
+                Logger.e("Local evaluation flags target cohorts but cohort targeting is not configured.")
```
Comment (tyiuhc) [inline src/main/kotlin/LocalEvaluationClient.kt:118]: nit: can be more explicit with logging during `evaluate`, e.g. log which flags do not have the required cohorts in storage

File: src/main/kotlin/RemoteEvaluationConfig.kt [hunk 6]
```diff
@@ -47,6 +51,16 @@ class RemoteEvaluationConfig internal constructor(
          */
         const val SERVER_URL = "https://api.lab.amplitude.com/"
```
Comment (tyiuhc) [inline src/main/kotlin/RemoteEvaluationConfig.kt:52]: should refer to US_SERVER_URL const?

File: src/main/kotlin/RemoteEvaluationConfig.kt [hunk 7]
```diff
@@ -47,6 +51,16 @@ class RemoteEvaluationConfig internal constructor(
...
 
+        /**
+         * "https://api.lab.amplitude.com/"
+         */
+        const val COHORT_SERVER_URL = "https://cohorts-v2.lab.amplitude.com/"
```
Comment (tyiuhc) [inline src/main/kotlin/RemoteEvaluationConfig.kt:57]: Should refer to US_COHORT_SERVER_URL const

File: src/main/kotlin/cohort/CohortStorage.kt [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+internal interface CohortStorage {
+    fun getCohort(cohortId: String): Cohort?
+    fun getCohorts(): Map<String, Cohort>
+    fun getCohortsForUser(userId: String, cohortIds: Set<String>): Set<String>
+    fun getCohortsForGroup(groupType: String, groupName: String, cohortIds: Set<String>): Set<String>
+    fun putCohort(cohort: Cohort)
+    fun deleteCohort(cohortId: String)
+}
+
+internal class ProxyCohortStorage(
```
Comment (tyiuhc) [inline src/main/kotlin/cohort/CohortStorage.kt:28]: where is this used?

File: src/main/kotlin/LocalEvaluationClient.kt (hunk 1 above)
Comment (bgiori) [reply src/main/kotlin/LocalEvaluationClient.kt:165]: You're right switching up logic

File: src/main/kotlin/LocalEvaluationClient.kt [hunk 9]
```diff
@@ -67,15 +102,115 @@ class LocalEvaluationClient internal constructor(
...
+            if (groupedCohortIds.isNotEmpty()) {
+                val flagKeys = flagConfigs.mapNotNull { flag ->
+                    val cohortIds = flag.getAllCohortIds()
+                    if (cohortIds.isEmpty()) {
+                        null
+                    } else {
+                        flag.key
+                    }
+                }
+                Logger.e("Local evaluation flags $flagKeys target cohorts but cohort targeting is not configured.")
```
Comment (tyiuhc) [inline src/main/kotlin/LocalEvaluationClient.kt:134]: during evaluation, should log warning for a flag if cohort targeting is configured, but cohort not in storage (due to e.g. download failure?)

**Review by tyiuhc — APPROVED** [review]

**Pull request merged**